player/js/utils/text/SkriptShaper.js

```diff
@@ -0,0 +1,999 @@
+/* eslint-disable max-classes-per-file */
+/* eslint-disable no-continue */
+/* eslint-disable no-plusplus */
+/* eslint-disable no-bitwise */
+
+const utils = {
+  css: function (element, property) {
+    return window.getComputedStyle(element, null).getPropertyValue(property);
+  },
+
+  compare: function (a, b) {
+    return ['top', 'left', 'right', 'bottom'].every((key) => a[key] === b[key]);
+  },
+
+  detectTextDirection: function (/* DOMRectList */rects) {
+    if (rects.length === 0) {
+      return TextDirection.Undefined;
+    } if (rects.length === 1) {
+      return TextDirection.LTR;
+    }
+    let left = rects[0].left;
+    let right = rects[0].right;
+    let goingRight;
+    let goingLeft;
+    for (const rect of rects) {
+      if (goingRight === undefined && goingLeft === undefined) {
+        // First rectangle
+        if (rects[0].left <= rects[rects.length - 1].right) {
+          goingLeft = false;
+          goingRight = true;
+        } else {
+          goingLeft = true;
+          goingRight = false;
+        }
+      } else if (utils.equal(right, rect.left)) { // We are going sequentially right
+        if (goingLeft) {
+          return TextDirection.Mixed;
+        }
+        console.assert(goingRight);
```

Comment:
can asserts be removed?

Reply:
I removed the assert but kept the logging since these are significant problems and I don't want them to go unnoticed at least for now when the code is not well tested.

Reply:
I updated all logging to be disabled (no-op) by default. We can enable for testing using a window flag. @bodymovin PTAL.